From 12a39bec00f66eb4eccaff1c34485bffeb4221d6 Mon Sep 17 00:00:00 2001
From: NLObP <38990417+NL0bP@users.noreply.github.com>
Date: Wed, 7 Aug 2024 16:57:52 +0300
Subject: [PATCH] [Game][Ai] Changes have been made to the UpdateAggroHelp
 method: (#1062)

* [Game][Ai] Changes have been made to the UpdateAggroHelp method:
- Now we use the AggroLinkSpecialRuleKind enumeration.

[Game] Added some enumerations:
- AggroLinkSpecialRuleKind;
- AggroTypes;
- NpcGroupAggroRuleKind;
- NpcSpawnType.

[Game][Ai] Eliminated spam messages "Failed to remove unit[{0}] aggro from NPC[{1}]".

[Packets] Disabled displaying information about the SCUnitModelPostureChangedPacket packet.

* Removed extra semicolon.

* [Game] Removed AggroTypes enums as a duplicate of an existing one.
---
 .../Core/Managers/UnitManagers/NpcManager.cs  |  2 +-
 .../Models/Game/AI/v2/Framework/Behavior.cs   | 62 ++++++++++++++-----
 .../Game/NPChar/AggroLinkSpecialRuleKind.cs   | 10 +++
 AAEmu.Game/Models/Game/NPChar/Npc.cs          | 26 ++++----
 .../Game/NPChar/NpcGroupAggroRuleKind.cs      |  8 +++
 AAEmu.Game/Models/Game/NPChar/NpcSpawnType.cs |  9 +++
 AAEmu.Game/Models/Game/NPChar/NpcTemplate.cs  |  2 +-
 7 files changed, 91 insertions(+), 28 deletions(-)
 create mode 100644 AAEmu.Game/Models/Game/NPChar/AggroLinkSpecialRuleKind.cs
 create mode 100644 AAEmu.Game/Models/Game/NPChar/NpcGroupAggroRuleKind.cs
 create mode 100644 AAEmu.Game/Models/Game/NPChar/NpcSpawnType.cs

diff --git a/AAEmu.Game/Core/Managers/UnitManagers/NpcManager.cs b/AAEmu.Game/Core/Managers/UnitManagers/NpcManager.cs
index b06511f78..c7ab0ea26 100644
--- a/AAEmu.Game/Core/Managers/UnitManagers/NpcManager.cs
+++ b/AAEmu.Game/Core/Managers/UnitManagers/NpcManager.cs
@@ -562,7 +562,7 @@ public void Load()
                         template.NpcAiParamId = reader.GetInt32("npc_ai_param_id");
                         template.NonPushableByActor = reader.GetBoolean("non_pushable_by_actor", true);
                         template.Banker = reader.GetBoolean("banker", true);
-                        template.AggroLinkSpecialRuleId = reader.GetInt32("aggro_link_special_rule_id");
+                        template.AggroLinkSpecialRuleId = (AggroLinkSpecialRuleKind)reader.GetInt32("aggro_link_special_rule_id");
                         template.AggroLinkHelpDist = reader.GetFloat("aggro_link_help_dist");
                         template.AggroLinkSightCheck = reader.GetBoolean("aggro_link_sight_check", true);
                         template.Expedition = reader.GetBoolean("expedition", true);
diff --git a/AAEmu.Game/Models/Game/AI/v2/Framework/Behavior.cs b/AAEmu.Game/Models/Game/AI/v2/Framework/Behavior.cs
index c20a2352a..bd5110dfa 100644
--- a/AAEmu.Game/Models/Game/AI/v2/Framework/Behavior.cs
+++ b/AAEmu.Game/Models/Game/AI/v2/Framework/Behavior.cs
@@ -1,10 +1,11 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Numerics;
+
 using AAEmu.Commons.Utils;
 using AAEmu.Game.Core.Managers;
 using AAEmu.Game.Core.Managers.World;
+using AAEmu.Game.Models.Game.Faction;
 using AAEmu.Game.Models.Game.Items.Templates;
 using AAEmu.Game.Models.Game.NPChar;
 using AAEmu.Game.Models.Game.Skills;
@@ -233,7 +234,7 @@ public bool CheckAggression()
             unitsWithDistance.Add((nearbyUnit, rangeOfUnit));
         }
         unitsWithDistance.Sort((p, q) => p.Item2.CompareTo(q.Item2));
-        
+
         foreach (var (unit, rangeOfUnit) in unitsWithDistance)
         {
             if (unit.IsDead || unit.Hp <= 0)
@@ -243,7 +244,7 @@ public bool CheckAggression()
             var maxHeightGap = Ai.Owner.CanFly ? (Ai.Owner.ModelSize * Ai.Owner.Scale * 3.5f) : (Ai.Owner.ModelSize * Ai.Owner.Scale * 1.5f);
 
             // Check if in front, and not too far up or down
-            if (MathUtil.IsFront(Ai.Owner, unit, Ai.Owner.Template.SightFovScale) && 
+            if (MathUtil.IsFront(Ai.Owner, unit, Ai.Owner.Template.SightFovScale) &&
                 Math.Abs(Ai.Owner.Transform.World.Position.Z - unit.Transform.World.Position.Z) < maxHeightGap)
             {
                 if (Ai.Owner.CanAttack(unit) && (rangeOfUnit < 1f || Ai.Owner.CanSeeTarget(unit)))
@@ -278,16 +279,16 @@ public void OnEnemyAlert(Unit target)
         //     var degree = MathUtil.ClampDegAngle(MathUtil.CalculateAngleFrom(Ai.Owner, player));
         //     player.SendMessage($"ObjId {Ai.Owner.ObjId} has seen you at a angle of {degree:F0}°");
         // }
-        
+
         // TODO: Tweak these values, or grab them from DB somewhere?
         Ai._alertEndTime = DateTime.UtcNow.AddSeconds(5);
         Ai._nextAlertCheckTime = DateTime.UtcNow.AddSeconds(7);
         // Ai.Owner.CurrentAggroTarget = target;
         Ai.Owner.SetTarget(target);
-        
+
         Ai.GoToAlert();
     }
-    
+
     public bool CheckAlert()
     {
         if (Ai._nextAlertCheckTime > DateTime.UtcNow)
@@ -299,7 +300,7 @@ public bool CheckAlert()
 
         var res = false;
         var nearbyUnits = WorldManager.GetAround<Unit>(Ai.Owner, Ai.Owner.Template.SightRangeScale * 15f);
-        
+
         // Sort by distance
         var unitsWithDistance = new List<(Unit, float)>();
         foreach (var nearbyUnit in nearbyUnits)
@@ -308,7 +309,7 @@ public bool CheckAlert()
             unitsWithDistance.Add((nearbyUnit, rangeOfUnit));
         }
         unitsWithDistance.Sort((p, q) => p.Item2.CompareTo(q.Item2));
-        
+
         foreach (var (unit, rangeOfUnit) in unitsWithDistance)
         {
             if (unit.IsDead || unit.Hp <= 0)
@@ -318,7 +319,7 @@ public bool CheckAlert()
             var maxHeightGap = Ai.Owner.CanFly ? (Ai.Owner.ModelSize * Ai.Owner.Scale * 4f) : (Ai.Owner.ModelSize * Ai.Owner.Scale * 1.75f);
 
             // Check if in front, and not too far up or down
-            if (MathUtil.IsFront(Ai.Owner, unit, Ai.Owner.Template.SightFovScale) && 
+            if (MathUtil.IsFront(Ai.Owner, unit, Ai.Owner.Template.SightFovScale) &&
                 Math.Abs(Ai.Owner.Transform.World.Position.Z - unit.Transform.World.Position.Z) < maxHeightGap)
             {
                 if (Ai.Owner.CanAttack(unit) && (rangeOfUnit < 1f || Ai.Owner.CanSeeTarget(unit)))
@@ -347,18 +348,49 @@ public bool CheckAlert()
         return res;
     }
 
-    public void UpdateAggroHelp(Unit abuser, int radius = 20)
+    public void UpdateAggroHelp(Unit abuser, int radius = 200)
     {
-        var npcs = WorldManager.GetAround<Npc>(Ai.Owner,  Ai.Owner.Template.AttackStartRangeScale * radius);
+        bool needHelp;
+        var npcs = WorldManager.GetAround<Npc>(Ai.Owner, Ai.Owner.Template.AttackStartRangeScale * radius);
         if (npcs == null)
+        {
             return;
+        }
 
         foreach (var npc in npcs
-                     .Where(npc => npc.Template.Aggression && !npc.IsInBattle && npc.Template.AcceptAggroLink)
+                     .Where(npc => !npc.IsInBattle && npc.Template.AcceptAggroLink)
                      .Where(npc => npc.GetDistanceTo(npc.Ai.Owner) <= npc.Template.AggroLinkHelpDist))
         {
-            if (Ai.Owner.IsInBattle)
-                return; // already in battle, let's not change the target
+            if (npc.Template.Aggression && npc.Template.AggroLinkSpecialRuleId == AggroLinkSpecialRuleKind.None)
+            {
+                needHelp = true;
+            }
+            else
+            {
+                if (!(npc.Template.AggroLinkSightCheck && npc.CanSeeTarget(abuser)))
+                {
+                    continue;
+                }
+
+                switch (npc.Template.AggroLinkSpecialRuleId)
+                {
+                    case AggroLinkSpecialRuleKind.FactionHelp when npc.Faction.Id == Ai.Owner.Faction.Id:
+                    case AggroLinkSpecialRuleKind.FriendlyHelp when npc.GetRelationStateTo(Ai.Owner) == RelationState.Friendly:
+                    case AggroLinkSpecialRuleKind.NeutralHelp when npc.GetRelationStateTo(Ai.Owner) == RelationState.Neutral:
+                    case AggroLinkSpecialRuleKind.EveryoneHelp:
+                        needHelp = true;
+                        break;
+                    case AggroLinkSpecialRuleKind.None:
+                    default:
+                        needHelp = false;
+                        break;
+                }
+            }
+
+            if (!needHelp)
+            {
+                continue;
+            }
 
             npc.Ai.Owner.AddUnitAggro(AggroKind.Damage, abuser, 1);
             npc.Ai.OnAggroTargetChanged();
@@ -400,7 +432,7 @@ public bool CheckFollowPath()
     {
         return Ai.PathHandler.HasPathMovementData();
     }
-    
+
     public Behavior SetDefaultBehavior()
     {
         Ai.SetDefaultBehavior(this);
diff --git a/AAEmu.Game/Models/Game/NPChar/AggroLinkSpecialRuleKind.cs b/AAEmu.Game/Models/Game/NPChar/AggroLinkSpecialRuleKind.cs
new file mode 100644
index 000000000..7c57e83ae
--- /dev/null
+++ b/AAEmu.Game/Models/Game/NPChar/AggroLinkSpecialRuleKind.cs
@@ -0,0 +1,10 @@
+﻿namespace AAEmu.Game.Models.Game.NPChar;
+
+public enum AggroLinkSpecialRuleKind : uint
+{
+    None = 0,
+    FactionHelp = 1,
+    FriendlyHelp = 2,
+    NeutralHelp = 3,
+    EveryoneHelp = 4
+}
diff --git a/AAEmu.Game/Models/Game/NPChar/Npc.cs b/AAEmu.Game/Models/Game/NPChar/Npc.cs
index 7f3dd3c8d..859987ec5 100644
--- a/AAEmu.Game/Models/Game/NPChar/Npc.cs
+++ b/AAEmu.Game/Models/Game/NPChar/Npc.cs
@@ -18,7 +18,6 @@
 using AAEmu.Game.Models.Game.Models;
 using AAEmu.Game.Models.Game.Skills;
 using AAEmu.Game.Models.Game.Skills.SkillControllers;
-using AAEmu.Game.Models.Game.Team;
 using AAEmu.Game.Models.Game.Units;
 using AAEmu.Game.Models.Game.Units.Movements;
 using AAEmu.Game.Models.Game.Units.Static;
@@ -44,7 +43,8 @@ public partial class Npc : Unit
     /// <summary>
     /// This is the "Idle Animation Id" that is used in UnitModelChangePosture, it can change depending on the time of the day
     /// </summary>
-    public uint AnimActionId {
+    public uint AnimActionId
+    {
         get
         {
             switch (Template.NpcPostureSets.Count)
@@ -64,7 +64,7 @@ public uint AnimActionId {
             }
         }
     }
-    
+
     public override float Scale => Template.Scale;
 
     public override byte RaceGender => (byte)(16 * Template.Gender + Template.Race);
@@ -82,7 +82,7 @@ public BaseUnit CurrentAggroTarget
 
             if (value != null)
                 SendPacketToPlayers([value], new SCAggroTargetChangedPacket(ObjId, value.ObjId));
-                // BroadcastPacket(new SCAggroTargetChangedPacket(ObjId, value.ObjId), false);
+            // BroadcastPacket(new SCAggroTargetChangedPacket(ObjId, value.ObjId), false);
 
             _currentAggroTarget = value;
         }
@@ -840,7 +840,7 @@ public override void DoDie(BaseUnit killer, KillReason killReason)
                     }
                 }
             }
-            
+
             foreach (var pl in eligiblePlayers)
             {
                 var plKillXP = 0;
@@ -883,7 +883,7 @@ public override void DoDie(BaseUnit killer, KillReason killReason)
                 }
 
                 //Now we need to scale XP based on level difference, which gets a bit more complex.
-               
+
 
                 if (pl.Level >= this.Level + 10 || pl.Level <= this.Level - 10)
                 {
@@ -948,7 +948,7 @@ private void ClearAllAggroTargetsAndCheckCombatState()
         }
         // Clear the aggro table
         AggroTable.Clear();
-        
+
         // Check if those target players still have aggro on something else, if not, clear their combat timers
         foreach (var player in playerAggroList)
         {
@@ -1031,7 +1031,7 @@ public void AddUnitAggro(AggroKind kind, Unit unit, int amount)
                 if (!player.Quests.IsQuestComplete(Template.EngageCombatGiveQuestId) && !player.Quests.HasQuest(Template.EngageCombatGiveQuestId))
                     player.Quests.AddQuest(Template.EngageCombatGiveQuestId);
             }
-            
+
             // Send initial hit packet as well
             unit.SendPacketToPlayers([this, unit], new SCCombatFirstHitPacket(this.ObjId, unit.ObjId, 0));
         }
@@ -1059,11 +1059,15 @@ public void ClearAggroOfUnit(Unit unit)
         {
             player.IsInAggroListOf.Remove(ObjId);
         }
-        
+
         // var player = unit as Character;
         // player?.SendMessage($"ClearAggroOfUnit {player.Name} for {this.ObjId}");
 
         var lastAggroCount = AggroTable.Count;
+        if (lastAggroCount <= 0)
+        {
+            return;
+        }
         if (AggroTable.TryRemove(unit.ObjId, out var value))
         {
             unit.Events.OnHealed -= OnAbuserHealed;
@@ -1243,7 +1247,7 @@ public void MoveTowards(Vector3 other, float distance, byte actorFlags = 4)
         moveType.RotationY = ry;
         moveType.RotationZ = rz;
         moveType.ActorFlags = actorFlags;     // 5-walk, 4-run, 3-stand still
-        moveType.Flags = MoveTypeFlags.Moving | (IsInBattle ? MoveTypeFlags.InCombat : 0);; // MoveTypeFlags.Stopping;
+        moveType.Flags = MoveTypeFlags.Moving | (IsInBattle ? MoveTypeFlags.InCombat : 0); // MoveTypeFlags.Stopping;
 
         moveType.DeltaMovement = new sbyte[3];
         moveType.DeltaMovement[0] = 0;
@@ -1281,7 +1285,7 @@ public void LookTowards(Vector3 other, byte flags = 4)
         moveType.RotationY = ry;
         moveType.RotationZ = rz;
         moveType.ActorFlags = flags;     // 5-walk, 4-run, 3-stand still
-        moveType.Flags = MoveTypeFlags.Moving | (IsInBattle ? MoveTypeFlags.InCombat : 0);; // 4;
+        moveType.Flags = MoveTypeFlags.Moving | (IsInBattle ? MoveTypeFlags.InCombat : 0); ; // 4;
 
         moveType.DeltaMovement = new sbyte[3];
         moveType.DeltaMovement[0] = 0;
diff --git a/AAEmu.Game/Models/Game/NPChar/NpcGroupAggroRuleKind.cs b/AAEmu.Game/Models/Game/NPChar/NpcGroupAggroRuleKind.cs
new file mode 100644
index 000000000..411479125
--- /dev/null
+++ b/AAEmu.Game/Models/Game/NPChar/NpcGroupAggroRuleKind.cs
@@ -0,0 +1,8 @@
+﻿namespace AAEmu.Game.Models.Game.NPChar;
+
+public enum NpcGroupAggroRuleKind : uint
+{
+    None = 0,
+    AggroLink = 1,
+    AggroShare = 2
+}
diff --git a/AAEmu.Game/Models/Game/NPChar/NpcSpawnType.cs b/AAEmu.Game/Models/Game/NPChar/NpcSpawnType.cs
new file mode 100644
index 000000000..7e9242721
--- /dev/null
+++ b/AAEmu.Game/Models/Game/NPChar/NpcSpawnType.cs
@@ -0,0 +1,9 @@
+﻿namespace AAEmu.Game.Models.Game.NPChar;
+
+public enum NpcSpawnType : uint
+{
+    OneNpc = 0,
+    Random = 1,
+    Precedence = 2,
+    Together = 3
+}
diff --git a/AAEmu.Game/Models/Game/NPChar/NpcTemplate.cs b/AAEmu.Game/Models/Game/NPChar/NpcTemplate.cs
index 19f5c2f41..ca1d3a54c 100644
--- a/AAEmu.Game/Models/Game/NPChar/NpcTemplate.cs
+++ b/AAEmu.Game/Models/Game/NPChar/NpcTemplate.cs
@@ -54,7 +54,7 @@ public class NpcTemplate
     public int NpcAiParamId { get; set; }
     public bool NonPushableByActor { get; set; }
     public bool Banker { get; set; }
-    public int AggroLinkSpecialRuleId { get; set; }
+    public AggroLinkSpecialRuleKind AggroLinkSpecialRuleId { get; set; }
     public float AggroLinkHelpDist { get; set; }
     public bool AggroLinkSightCheck { get; set; }
     public bool Expedition { get; set; }
